### PR TITLE
Slayer Skill

### PR DESCRIPTION
**Pull request by themrrobert**
### Description:
Slayer skill

### Other checks:

-   [x] I have tested all my changes thoroughly.

### CODE REVIEW

File: src/commands/Minion/autoslay.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+export default class extends BotCommand {
+	public constructor(store: CommandStore, file: string[], directory: string) {
+		super(store, file, directory, {
+			altProtection: true,
+			oneAtTime: true,
+			cooldown: 1,
+			usage: '[_mode:...string]',
+			aliases: ['as', 'slay'],
+			usageDelim: ' ',
+			description: 'Sends your minion to kill your slayer monster monsters.'
```
Comment (gc) [inline src/commands/Minion/autoslay.ts:18]: grammatical error in description

File: src/commands/Minion/autoslay.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
+import { CommandStore, KlasaMessage } from 'klasa';
+
+import killableMonsters from '../../lib/minions/data/killableMonsters';
+import { minionNotBusy, requiresMinion } from '../../lib/minions/decorators';
+import { UserSettings } from '../../lib/settings/types/UserSettings';
+import { AutoslayOptionsEnum, getUsersCurrentSlayerInfo } from '../../lib/slayer/slayerUtil';
+import { BotCommand } from '../../lib/structures/BotCommand';
+
+export default class extends BotCommand {
```
Comment (gc) [inline src/commands/Minion/autoslay.ts:9]: command has no examples

File: src/commands/Minion/fletch.ts [hunk 3]
```diff
@@ -63,6 +66,22 @@ export default class extends BotCommand {
 			);
 		}
 
+		if (fletchable.requiredSlayerUnlocks) {
+			let mySlayerUnlocks = msg.author.settings.get(UserSettings.Slayer.SlayerUnlocks);
+			let mySlayerUnlocksIter: SlayerTaskUnlocksEnum[] = [];
+
+			mySlayerUnlocks.forEach(msu => {
+				mySlayerUnlocksIter.push(msu as SlayerTaskUnlocksEnum);
+			});
```
Comment (gc) [inline src/commands/Minion/fletch.ts:75]: this `forEach` seems totally useless, whats going on here?

File: src/commands/Minion/k.ts [hunk 4]
```diff
@@ -110,19 +135,75 @@ export default class extends BotCommand {
...
+
 		const maxTripLength = msg.author.maxTripLength(Activity.MonsterKilling);
 
 		// If no quantity provided, set it to the max.
 		if (quantity === null) {
 			quantity = floor(maxTripLength / timeToFinish);
 		}
+		if (typeof quantity !== 'number') quantity = parseInt(quantity);
+		if (isOnTask) {
+			// Todo: Probably handle this in a separate function to make everything easier.
```
Comment (gc) [inline src/commands/Minion/k.ts:185]: All todos should be capitalized `// TODO: <text>`
todos should be addressed in your PR if possible, although not totally required.

File: src/commands/Minion/k.ts [hunk 5]
```diff
@@ -176,6 +257,21 @@ export default class extends BotCommand {
 			duration *= 0.9;
 		}
 
+		// Todo add a new 'recurring cost' field to the KillableMonster that loops and does this.
+		// Needs some thinking, because it needs to have 1 per qty, and then timed ones.
+		// Remove antidote++(4) from hydras + alchemical hydra
```
Comment (gc) [inline src/commands/Minion/k.ts:262]: Keep the todo in 1 line, and formatted like mentioned above.

File: src/commands/Minion/offer.ts [hunk 6]
```diff
@@ -52,6 +63,41 @@ export default class extends BotCommand {
 			quantity = null;
 		}
 
+		if (boneName.toLowerCase() === 'unsired') {
+			const unsiredsOwned = userBank.amount(itemID('Unsired'));
```
Comment (gc) [inline src/commands/Minion/offer.ts:67]: you dont need `itemID` here

File: src/commands/Minion/slayershop.ts [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+
+	@requiresMinion
+	async run(msg: KlasaMessage, [_input]: [string]) {
+		if (
+			msg.flagArgs.unlocks ||
+			msg.flagArgs.help ||
+			_input === 'help' ||
+			_input === 'unlocks'
+		) {
+			// const myUnlocks = SlayerRewardsShop.filter(srs => { return srs.item !== undefined });
```
Comment (gc) [inline src/commands/Minion/slayershop.ts:36]: random comment here.

File: src/commands/Minion/slayershop.ts [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+			return msg.channel.sendFile(Buffer.from(unlockTable), `slayerRewardsItems.txt`);
+		}
+
+		let unlocks: string[] = [];
+		const myUnlocks = await msg.author.settings.get(UserSettings.Slayer.SlayerUnlocks);
+		const myPoints = await msg.author.settings.get(UserSettings.Slayer.SlayerPoints);
+
+		myUnlocks.forEach(u => {
+			unlocks.push(getSlayerReward(u));
+		});
```
Comment (gc) [inline src/commands/Minion/slayershop.ts:59]: another weird `forEach`. you should always try to use `.map` intead of forEach.

File: src/commands/Minion/slayershop.ts [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+			});
+			const unlockTable = table([
+				['Slayer Points', 'Name', 'Description', 'Type'],
+				...myUnlocks.map(i => [i.slayerPointCost, i.name, i.desc, 'item/buy'])
+			]);
+			return msg.channel.sendFile(Buffer.from(unlockTable), `slayerRewardsItems.txt`);
+		}
+
+		if (buyableName === '') {
+			throw `You must specify an item to purchase.\nTry:\n\`${msg.cmdPrefix}sls buy --help\``;
```
Comment (gc) [inline src/commands/Minion/slayershop.ts:84]: we no longer throw strings like this in commands, you should do `return msg.channel.send()`

File: src/commands/Minion/slayertask.ts [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+export default class extends BotCommand {
+	public constructor(store: CommandStore, file: string[], directory: string) {
+		super(store, file, directory, {
+			oneAtTime: true,
+			cooldown: 2,
+			altProtection: true,
+			categoryFlags: ['minion'],
+			aliases: ['st'],
+			description: 'slayer',
+			examples: ['+skillcape mining'],
```
Comment (gc) [inline src/commands/Minion/slayertask.ts:27]: add examples and a description

File: src/commands/Minion/slayertask.ts [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+			let mobs: string[] = [];
+			let outstr =
+				`You have a maximum of ${maxBlocks} task blocks. You are using ${myBlockList.length}` +
+				` and have ${maxBlocks - myBlockList.length} remaining\n\nBlocked Tasks:\n`;
+			const myBlockedMonsters = Monsters.filter(m => {
+				return myBlockList.includes(m.id);
+			});
+			myBlockedMonsters.forEach(m => {
+				mobs.push(`${m.id}: ${getCommonTaskName(m)}`);
+			});
```
Comment (gc) [inline src/commands/Minion/slayertask.ts:59]: another weird `forEach`, just do something like `let mobs = something.map(m => ``);`

File: src/commands/Minion/slayertask.ts [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+					.find(m => m.aliases.some(alias => stringMatches(alias, input))) ?? null
+			: rememberedSlayerMaster !== ''
+			? slayerMasters
+					.filter(m => userCanUseMaster(msg.author, m))
+					.find(m =>
+						m.aliases.some(alias => stringMatches(alias, rememberedSlayerMaster))
+					) ?? null
+			: null;
+
+		// TODO: Delete this
```
Comment (gc) [inline src/commands/Minion/slayertask.ts:184]: todo here

File: src/commands/Minion/slayertask.ts [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+		msg.author.log(`Remembered: ${rememberedSlayerMaster}`);
+		msg.author.log(`slayermaster: ${slayerMaster?.name}`);
+
+		const matchedSlayerMaster = input
+			? slayerMasters.find(m => m.aliases.some(alias => stringMatches(alias, input))) ?? null
+			: null;
+
+		// Special handling for Turael skip
+		if (currentTask && input && slayerMaster && slayerMaster.name === 'Turael') {
+			// TODO: Make sure they aren't already on a Turael task.
```
Comment (gc) [inline src/commands/Minion/slayertask.ts:194]: todo here

File: src/lib/data/buyables/slayerBuyables.ts [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+		}),
+		gpCost: 50_000
+	},
+	{
+		name: 'Rock hammer',
+		outputItems: resolveNameBank({
+			'Rock hammer': 1
+		}),
+		gpCost: 50_000
+	},
```
Comment (gc) [inline src/lib/data/buyables/slayerBuyables.ts:82]: Some of these items are never used?

File: src/lib/slayer/slayerUtil.ts [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+		MonsterSlayerMaster.Krystilia
+	];
+	return osjsSlayerMaster[slayerMasterID];
+}
+
+export function getSlayerReward(id: SlayerTaskUnlocksEnum): string {
+	let name = '';
+	SlayerRewardsShop.forEach(srs => {
+		if (srs.id === id) name = srs.name;
+	});
```
Comment (gc) [inline src/lib/slayer/slayerUtil.ts:247]: another weird forEach

**Review by gc — CHANGES_REQUESTED** [review]

File: src/commands/Minion/autoslay.ts (hunk 1 above)
Comment (themrrobert) [reply src/commands/Minion/autoslay.ts:18]: resolved

File: src/lib/skilling/skills/crafting/craftables/built.ts [hunk 16]
```diff
@@ -11,21 +11,6 @@ const Built: Craftable[] = [
...
-		xp: 0,
-		inputItems: resolveNameBank({
-			'Black mask': 1,
-			Earmuffs: 1,
-			Facemask: 1,
-			'Nose peg': 1,
-			'Spiny helmet': 1,
-			'Enchanted gem': 1
-		}),
-		tickRate: 3
```
Comment (gc) [inline src/lib/skilling/skills/crafting/craftables/built.ts:27]: This removes the ability to craft a slayer hem? im confused

**Review by gc — CHANGES_REQUESTED** [review]

File: src/lib/skilling/skills/crafting/craftables/built.ts (hunk 16 above)
Comment (themrrobert) [reply src/lib/skilling/skills/crafting/craftables/built.ts:27]: slayer helm isn't crafted, it's created (instantly)

but i did forget the 55 crafting requirement, i'll add that to the creatable.

File: src/commands/Owner/rp.ts [hunk 17]
```diff
@@ -27,6 +30,21 @@ export default class extends BotCommand {
...
+				if (!currentTask) return msg.channel.send(`You have no task.`);
+				await currentTask.remove();
+				return msg.channel.send(`Deleted your task.`);
+			}
+			case 'slayerreset': {
+				await SlayerTaskTable.delete({ user: await getNewUser(msg.author.id) });
+				await msg.author.settings.reset(UserSettings.Skills.Slayer);
+				return msg.channel.send(`Deleted ALL your task history, your slayer XP`);
+			}
+		}
```
Comment (gc) [inline src/commands/Owner/rp.ts:46]: This would be usable in the live bot! Time to delete these

**Review by gc — CHANGES_REQUESTED** [review]

File: src/commands/Owner/rp.ts (hunk 17 above)
Comment (Alexsuperfly) [reply src/commands/Owner/rp.ts]: or just put them in hax if youd want them for later

File: src/lib/data/cox.ts [hunk 18]
```diff
@@ -347,6 +347,12 @@ const itemBoosts = [
 			item: getOSItem('Bandos godsword (or)'),
 			boost: 2.5
 		}
+	],
+	[
+		{
+			item: getOSItem('Dragon hunter lance'),
```
Comment (J4stMart) [inline src/lib/data/cox.ts:353]: Applying the boost like this results in a 16.7% speed upgrade as supposed the likely intended 5%. For cox to allow more boost, the boost system should be rewritten. 
As many things in slayer won't be finished before it is being released. I would propose to exclude this from the release of slayer as well, until a better cox system has been made

**Pull request merged**